nixos/modules/services/web-apps/hedgedoc.nix

```diff
@@ -1018,58 +271,99 @@ in
-          cfg.settings.dbURL != "" && cfg.settings.dbURL != null
-        );
-        message = "Database configuration for HedgeDoc missing."; }
-    ];
-    users.groups.${name} = {};
+    users.groups.${name} = { };
     users.users.${name} = {
       description = "HedgeDoc service user";
       group = name;
-      extraGroups = cfg.groups;
-      home = cfg.workDir;
-      createHome = true;
       isSystemUser = true;
     };
 
-    systemd.services.hedgedoc = {
-      description = "HedgeDoc Service";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "networking.target" ];
-      preStart = ''
-        ${pkgs.envsubst}/bin/envsubst \
-          -o ${cfg.workDir}/config.json \
-          -i ${prettyJSON cfg.settings}
-        mkdir -p ${cfg.settings.uploadsPath}
-      '';
-      serviceConfig = {
-        WorkingDirectory = cfg.workDir;
-        StateDirectory = [ cfg.workDir cfg.settings.uploadsPath ];
-        ExecStart = "${cfg.package}/bin/hedgedoc";
-        EnvironmentFile = mkIf (cfg.environmentFile != null) [ cfg.environmentFile ];
-        Environment = [
-          "CMD_CONFIG_FILE=${cfg.workDir}/config.json"
-          "NODE_ENV=production"
-        ];
-        Restart = "always";
-        User = name;
-        PrivateTmp = true;
-      };
+    services.hedgedoc.settings = {
+      defaultNotePath = mkDefault "${cfg.package}/public/default.md";
+      docsPath = mkDefault "${cfg.package}/public/docs";
+      viewPath = mkDefault "${cfg.package}/public/views";
     };
+
+    systemd.services.hedgedoc =
+      let
+        # Some settings that we set as null will break hedgedoc,
+        # as it expects them either to be undefined or set.
+        cleanConfig = pkgs.runCommandLocal "hedgedoc-config.json"
+          {
+            nativeBuildInputs = [ pkgs.jq ];
+          } ''
+          jq '{production:del(.[]|nulls)|del(.[][]?|nulls)}' \
+            < ${settingsFormat.generate "hedgedoc-ugly.json" cfg.settings} \
+            > $out
+        '';
+      in
+      {
+        description = "HedgeDoc Service";
+        documentation = [ "https://docs.hedgedoc.org/" ];
+        wantedBy = [ "multi-user.target" ];
+        after = [ "networking.target" ];
+        preStart = ''
+          ${pkgs.envsubst}/bin/envsubst \
+            -o /var/lib/${name}/config.json \
+            -i ${cleanConfig}
+          ${pkgs.coreutils}/bin/mkdir -p ${cfg.settings.uploadsPath}
+        '';
+        serviceConfig = {
+          User = name;
+          Group = name;
+
+          Restart = "always";
+          ExecStart = "${cfg.package}/bin/hedgedoc";
+          RuntimeDirectory = [ name ];
+          StateDirectory = [ name ];
+          WorkingDirectory = "/var/lib/${name}";
+          ReadOnlyPaths = builtins.filter (x: x != null) (with cfg.settings; [
+            defaultNotePath
+            dhParamPath
+            docsPath
+            sslCertPath
+            sslKeyPath
+            viewPath
+          ] ++ sslCAPath);
```

Comment:
I am not a fan of setting this as it is a huge pain to debug when things go wrong. With the *Directories above we should already better than nothing

Reply:
Isn't it worse for the normal user to debug this when they set their ssl keys to be somewhere other than `/var/lib/hedgedoc`? With all the new hardening, not doing something equivalent to this would be a breaking change for users who have set their stuff to be outside `StateDirectory`.

Reply:
We could also just remove ReadOnlyPaths again and be slightly less hardened but also things would be simpler.

Reply:
Sure. That resolves the two other comments